### PR TITLE
Support new controller options

### PR DESCRIPTION
**Pull request by chrismccord**
Ready for early review, but not ready for merge. Start of #4975

### CODE REVIEW

File: priv/templates/phx.gen.json/fallback_controller.ex [hunk 1]
```diff
@@ -10,15 +10,15 @@ defmodule <%= inspect context.web_module %>.FallbackController do
   def call(conn, {:error, %Ecto.Changeset{} = changeset}) do
     conn
     |> put_status(:unprocessable_entity)
-    |> put_view(<%= inspect context.web_module %>.ChangesetView)
+    |> put_view(html: <%= inspect context.web_module %>.ChangesetView, json: <%= inspect context.web_module %>.ChangesetView)
```
Comment (chrismccord) [inline priv/templates/phx.gen.json/fallback_controller.ex:13]: need to bump these once we're ready elsewhere

File: lib/phoenix/controller.ex [hunk 2]
```diff
@@ -757,15 +830,18 @@ defmodule Phoenix.Controller do
   def render(conn, view, template, assigns)
       when is_atom(view) and (is_binary(template) or is_atom(template)) do
     conn
-    |> put_view(view)
+    |> put_view(any: view)
```
Comment (josevalim) [inline lib/phoenix/controller.ex:833]: Bug?

File: lib/phoenix/controller.ex [hunk 3]
```diff
@@ -757,15 +830,18 @@ defmodule Phoenix.Controller do
...
   end
 
   defp render_and_send(conn, format, template, assigns) do
     template = template_name(template, format)
     view =
-      Map.get(conn.private, :phoenix_view) ||
-        raise "a view module was not specified, set one with put_view/2"
+      case Map.get(conn.private, :phoenix_view) do
+        %{^format => view} -> view
+        %{:_ => view} when is_atom(view) and not is_nil(view) -> view
```
Comment (josevalim) [inline lib/phoenix/controller.ex:842]: Shouldn't this come first?

File: lib/phoenix/controller.ex [hunk 4]
```diff
@@ -1589,6 +1669,26 @@ defmodule Phoenix.Controller do
...
+
+          other ->
+            raise ArgumentError, """
+            expected :layouts to be a list of format module pairs of the form: [html: LayoutHTML] or [html: {LayoutHTML, :app}]
+
+            Got: #{inspect(other)}
+            """
+        end)
+
+      :error ->
```
Comment (josevalim) [inline lib/phoenix/controller.ex:1690]: Move the whole previous code to this branch?

File: priv/templates/phx.gen.json/fallback_controller.ex (hunk 1 above)
Comment (josevalim) [inline priv/templates/phx.gen.json/fallback_controller.ex:13]: I think those changes are not necessary given our new approach that `:_` is always first?

File: priv/templates/phx.gen.json/fallback_controller.ex [hunk 5]
```diff
@@ -10,15 +10,15 @@ defmodule <%= inspect context.web_module %>.FallbackController do
...
     |> render("error.json", changeset: changeset)
   end
 
   <% end %># This clause is an example of how to handle resources that cannot be found.
   def call(conn, {:error, :not_found}) do
     conn
     |> put_status(:not_found)
-    |> put_view(<%= inspect context.web_module %>.ErrorView)
+    |> put_view(html: <%= inspect context.web_module %>.ErrorView, json: <%= inspect context.web_module %>.ErrorView)
```
Comment (josevalim) [inline priv/templates/phx.gen.json/fallback_controller.ex:21]: I think those changes are not necessary given our new approach that `:_` is always first?

File: lib/phoenix/endpoint/render_errors.ex [hunk 6]
```diff
@@ -105,15 +105,14 @@ defmodule Phoenix.Endpoint.RenderErrors do
...
 
-    reason = Exception.normalize(kind, reason, stack)
     format = Controller.get_format(conn)
+    reason = Exception.normalize(kind, reason, stack)
     template = "#{conn.status}.#{format}"
     assigns = %{kind: kind, reason: reason, stack: stack, status: conn.status}
 
     conn
-    |> Controller.put_view(view)
+    |> Controller.put_view([{format, view}])
```
Comment (josevalim) [inline lib/phoenix/endpoint/render_errors.ex:115]: I think those changes are not necessary given our new approach that `:_` is always first?

File: lib/phoenix/controller.ex [hunk 7]
```diff
@@ -433,22 +433,72 @@ defmodule Phoenix.Controller do
...
+
+      %{^format => value} ->
+        value
+
+      value when value in [nil, false] ->
+        false
+
+      %{} = formats when map_size(formats) == 1 ->
+        {_format, value} = (Enum.at(formats, 0))
+        value
```
Comment (josevalim) [inline lib/phoenix/controller.ex:480]: I think no longer necessary if `:_`is always first?

```suggestion
```

File: lib/phoenix/controller.ex [hunk 8]
```diff
@@ -492,26 +544,46 @@ defmodule Phoenix.Controller do
...
-  defp do_put_layout(conn, private_key, false) do
-    put_private(conn, private_key, false)
-  end
+  defp do_put_layout(conn, private_key, kind, layouts) when is_list(layouts) do
+    formats =
+      Enum.map(layouts, fn
+        {format, false} ->
+          {format, false}
+
+        {format, {mod, layout}} when is_atom(mod) and (is_atom(layout) or is_binary(layout)) ->
```
Comment (josevalim) [inline lib/phoenix/controller.ex:559]: Let's already deprecate binary layouts?

File: lib/phoenix/controller.ex (hunk 7 above)
Comment (chrismccord) [reply lib/phoenix/controller.ex:480]: we still need this check to avoid the `get_format` lookup if there is only a single one

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**